File: 160-intersection-of-two-linked-lists/160-intersection-of-two-linked-lists.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function getIntersectionNode(headA: ListNode | null, headB: ListNode | null): ListNode | null {
    let currentA = headA;
    let currentB = headB;
    const list = new Set();
    
    while(currentA || currentB ) {
        
        if(currentB && list.has(currentB)) {
           return currentB;
        } 
        if(currentB){
            list.add(currentB);
        }
        
        if(currentA && list.has(currentA)) {
           return currentA;
        }
        if(currentA) {
            list.add(currentA);
        }
        //console.log(list);
        currentA = currentA?.next;
        currentB = currentB?.next
    }
        //console.log(list);
    return null;
    
};
    
    
    // 10 2 
    // 1 3 4 5 6 7 8 9 2

File: 160-intersection-of-two-linked-lists/160-intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        list = set()
        head1, head2 = headA, headB
        
        while head1 or head2:
            if head1 in list:
                return head1
            if head1 is not None:
                list.add(head1)
                
            if head2 in list:
                return head2
            if head2 is not None:
                list.add(head2)
            
            head1 = head1.next if head1 else None
            head2 = head2.next if head2 else None
        return None
        